monasca-alarms/monasca_alarm_definition.py

```diff
@@ -541,6 +518,71 @@ def _env(*vars, **kwargs):
     return kwargs.get('default', '')
 
 
+def get_keystone_client(keystone_args):
+    auth = Password(**keystone_args)
+
+    for i in range(KEYSTONE_RETRIES):
+        try:
+            session = Session(auth=auth,
+                              app_name='monasca-alarms',
+                              user_agent='monasca-alarms',
+                              timeout=KEYSTONE_TIMEOUT,
+                              verify=KEYSTONE_VERIFY,
+                              cert=KEYSTONE_CERT)
+
+            # we will end up throwing away the client we make here since
+            # somehow there is no way to pass a working Keystone client to the
+            # monascaclient (thanks osc-lib)
+            discover = Discover(session=session)
+            discover.create_client()
+            return session
+        except (RetriableConnectionFailure, Unauthorized) as ex:
+            print('Keystone is not yet ready (attempt {} of {}): {}'.format(
+                i, KEYSTONE_RETRIES, ex.message
+            ))
+
+            if i < KEYSTONE_RETRIES - 1:
+                time.sleep(KEYSTONE_DELAY)
+
+        except Exception as ex:
+            print('Unexpected error while connecting to Keystone, giving up: ',
+                  ex.message)
+            raise
+
+    print('Could not connect to Keystone after {} retries, '
+          'giving up!'.format(KEYSTONE_RETRIES))
+    raise Exception('could not connect to keystone')
+
+
+def get_monasca_client(args, keystone_args):
+    api_url = args.get('monasca_api_url')
+    if not api_url:
+        raise Exception('Error: monasca_api_url is required')
+
+    monasca = client.Client(args['api_version'], api_url, **keystone_args)
+    print('Using Monasca at {}'.format(api_url))
+
+    for i in range(MONASCA_RETRIES):
+        try:
+            # just make an api call that should always succeed
+            monasca.notifications.list(limit=1)
```

Comment:
I'm still confused about why we can't just let the monascaclient authenticate to keystone. It seems like get_keystone_client does a whole bunch of work that is just repeated in the monascaclient. Why do we need to authenticate to keystone explicitly?

Reply:
Yeah, originally I was hoping not to touch the `MonascaLoadDefinition` class at all, so it made more sense to have that in a separate step. Since I had to do that anyway, they probably could be combined now.